ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java

```diff
@@ -397,14 +387,95 @@ protected AstNode visitJoin(final Join node, final Void context) {
       return null;
     }
 
-    private boolean validJoin(
-        final SourceName leftName,
-        final SourceName rightName,
+    private void throwOnSelfJoin(final AliasedDataSource left, final AliasedDataSource right) {
+      if (left.getDataSource().getName().equals(right.getDataSource().getName())) {
+        throw new KsqlException(
+            "Can not join '" + left.getDataSource().getName().toString(FormatOptions.noEscape())
+                + "' to '" + right.getDataSource().getName().toString(FormatOptions.noEscape())
+                + "': self joins are not yet supported."
+        );
+      }
+    }
+
+    private void throwOnIncompleteJoinCriteria(
+        final AliasedDataSource left,
+        final AliasedDataSource right,
         final SourceName leftExpressionSource,
         final SourceName rightExpressionSource
     ) {
-      return ImmutableSet.of(leftExpressionSource, rightExpressionSource)
-          .containsAll(ImmutableList.of(leftName, rightName));
+      final boolean valid = ImmutableSet.of(leftExpressionSource, rightExpressionSource)
+          .containsAll(ImmutableList.of(left.getAlias(), right.getAlias()));
+
+      if (!valid) {
+        throw new KsqlException(
+            "Each side of the join must reference exactly one source and not the same source. "
+                + "Left side references " + leftExpressionSource
+                + " and right references " + rightExpressionSource
+        );
+      }
+    }
+
+    private void throwOnIncompatibleSourceWindowing(
+        final AliasedDataSource left,
+        final AliasedDataSource right
+    ) {
+      final Optional<WindowType> leftWindowType = left.getDataSource()
+          .getKsqlTopic()
+          .getKeyFormat()
+          .getWindowInfo()
+          .map(WindowInfo::getType);
+
+      final Optional<WindowType> rightWindowType = right.getDataSource()
+          .getKsqlTopic()
+          .getKeyFormat()
+          .getWindowInfo()
+          .map(WindowInfo::getType);
+
+      if (leftWindowType.isPresent() != rightWindowType.isPresent()) {
+        throw windowedNonWindowedJoinException(left, right, leftWindowType, rightWindowType);
+      }
+
+      if (!leftWindowType.isPresent()) {
+        return;
+      }
+
+      final WindowType leftWt = leftWindowType.get();
```

Comment:
Just to make sure: Session can only join with session whereas hopping and tumbling can join with each other interchangeably?

Reply:
Nevermind, tests are self-explanatory :)